File: leetcode/python/0153. Find Minimum in Rotated Sorted Array.py
# Time complexity: O(logN)
# Space complexity: O(1)

class Solution(object):
    def findMin(self, nums):
        return nums[self.find_pivot(nums)]
    
    def find_pivot(self, nums):
        left = 0
        right = len(nums) - 1

        while right >= left:
            mid = (left + right) / 2

            if nums[mid] < nums[mid-1]:
                return mid
            elif nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid  - 1
        return -1

File: leetcode/java/0153. Find Minimum in Rotated Sorted Array.java
package leetcode.java;

// Time complexity: O(logN)
// Space complexity: O(1)

class Solution {
    public int findMin(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        if (nums.length == 1) {
            return nums[0];
        }

        while (right >= left) {
            int mid = (right + left) / 2;
            if (nums[mid] < nums[mid-1 == -1 ? right : mid-1]) {
                return nums[mid];
            } else if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return nums[0];
    }
}